crates/wasmtime/src/component/component.rs

```diff
@@ -239,7 +243,24 @@ impl Component {
             }),
         );
         let unlinked_compile_outputs = compile_inputs.compile(&engine)?;
-        let types = types.finish();
+
+        let (types, ty) = types.finish(
+            component
+                .component
+                .import_types
+                .iter()
+                .map(|(_, (name, ty))| (name.clone(), *ty)),
+            component.component.exports.iter().filter_map(|(name, ty)| {
+                let ty = match ty {
+                    Export::LiftedFunction { ty, .. } => TypeDef::ComponentFunc(*ty),
+                    Export::ModuleImport { ty, .. } => TypeDef::Module(*ty),
+                    Export::Instance { ty: Some(ty), .. } => TypeDef::ComponentInstance(*ty),
+                    Export::Type(ty) => *ty,
+                    _ => return None,
```

Comment:
This is the main thing that I'm worried about with this PR right now. This means that if a component exports a locally-defined core wasm module and/or a "bag of items" instance then wasmtime will lie and not actually report the export from the type. I think it should be possible to implement this though with some more type-threading-around, however. All the types of all intermediate values are present and accounted for during validation, we just need to pluck it out at the right time and thread it through the system to show up here.

Reply:
It appears that we can pluck most of these out, but not for all cases, see updated implementation